src/components/Dropdown/index.js

```diff
@@ -0,0 +1,186 @@
+  <div class="dropdown" data-select="dropdown-container">
+    <div class="dropdown__toggle" data-select="dropdown-toggle">
+      <span class="dropdown__selected dropdown__selected--label" data-select="dropdown-selected"></span>
+      <span class="dropdown__icon" data-select="dropdown-icon">
+        <svg fill="currentColor" stroke-width="0" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512" style="overflow: visible; color: currentcolor;" height="1em" width="1em">
+          <path d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"></path>
+        </svg>
+      </span>
+    </div>
+    <ul class="dropdown__options" data-select="dropdown-options"></ul>
+  </div>
+`;
+
+export default function Dropdown({
+  items = [],
+  placeholder = 'Select an option',
+  value = null,
+  cssClass = '',
+} = {}) {
+  Component.call(this, { html, events });
+
+  this.placeholder = placeholder;
+  this.items = new Map();
+  this.itemSelected = null;
+
+  this.onSelect = (item) => {
+    const existsAndIsDifferent = this.itemSelected != null && this.itemSelected !== item;
+    if (existsAndIsDifferent) this.itemSelected.unselect();
+
+    this.itemSelected = item;
+    this.setValue(item.getValue());
+    this.emit('select', item);
+
+    if (this.isOpen()) this.close();
+  };
+
+  this.onUnselect = (item) => {
+    const isDifferent = this.itemSelected !== item;
+    if (isDifferent) return;
+
+    this.itemSelected = null;
+    this.setValue(null);
+    this.emit('unselect', item);
+  };
+
+  items.map((item) => this.addItem(item));
+  this.setValue(value);
+
+  this.selected
+    .get('dropdown-toggle')
+    .addEventListener('click', () => this.toggle());
+
+  const closeOnClickOutside = (event) => {
+    const isOutside = !event
+      .composedPath()
+      .includes(this.selected.get('dropdown-container'));
+
+    if (!isOutside) return;
+
+    if (this.isOpen()) this.close();
+  };
+
+  this.listen('mount', () => document.addEventListener('click', closeOnClickOutside),
+  );
+  this.listen('unmount', () => document.removeEventListener('click', closeOnClickOutside),
+  );
+
+  if (cssClass) {
+    this.selected.get('dropdown-container').classList.add(cssClass);
+  }
+}
+
+Dropdown.prototype = Object.assign(Dropdown.prototype, Component.prototype, {
+  addItem(props = {}) {
+    if (this.items.has(props.value)) return;
+
+    const item = new DropdownItem(props);
+    const $list = this.selected.get('dropdown-options');
+    item.mount($list);
+    item.listen('select', this.onSelect);
+    item.listen('unselect', this.onUnselect);
+    this.items.set(props.value, item);
+    this.emit('item:add', item);
+  },
+
+  removeItem(value = '') {
+    if (!this.items.has(value)) return;
+
+    const item = this.items.get(value);
+    item.unmount();
+    this.items.delete(value);
+
+    if (item.getValue() === this.getValue()) {
```

Comment:
isso aqui não deveria ser `this.itemSelected.getValue() === value`?

Porque vc tem a referencia do item atual e tambem tem essa função de `getValue()`? Acho que seria bem mais facil trabalhar apenas com o `selectedItem` (renomeando), do jeito que ta ficou bem confuso

Reply:
se fizer `this.itemSelected.getValue() === value` da erro que o getValue não pode ler valor nulo.